### ISSUE TITLE
[attention] the non-split-K sink reduction is uncovered: sinks can be dropped from the softmax denominator undetected

### ISSUE THREAD

**Issue by kekzl** [kekzl/imp#1303]
## Summary

The attention-sink term can be deleted from the non-split-K decode reduction and every test still passes: the only sink tests are deliberately shaped so that **split-K** fires, and split-K uses a different reduction function.

## Repro

Commit: `ad067d76`
Mutant `M31`, `src/compute/attention_paged_common.cuh` in `crosswarp_reduce_and_write`:

```diff
         float global_l = 0.0f;
         for (int w = 0; w < NUM_WARPS; w++)
             global_l += expf(warp_max[w] - global_max) * warp_l[w];
-        if (attn_sinks)
-            global_l += expf(__half2float(attn_sinks[head_idx]) - global_max);
```

Command: `tools/mutation/run.py --only M31`
Environment: RTX 5090, CUDA 13.3.1, `build-dev` Release, full local suite (8 binaries, all `IMP_TEST_MODEL*` set).

## Expected vs Actual

Expected: `PagedAttentionTest.GQA_SplitK_HD64_Sinks` fails — it is the sink test, and it has a proper CPU reference that computes the sink column.
Actual: **SURVIVED** the full suite, and 7/7 green on a re-check against an isolated `DetEvalE2ETest.*:PrefixCacheE2ETest.*` oracle.

## Evidence

`tests/test_paged_attention.cu:799`:

```cpp
TEST(PagedAttentionTest, GQA_SplitK_HD64)       { run_gptoss_shape_splitk_case(/*with_sinks=*/false); }
TEST(PagedAttentionTest, GQA_SplitK_HD64_Sinks) { run_gptoss_shape_splitk_case(/*with_sinks=*/true);  }
```

and the shape those tests use (`:681`):

```cpp
constexpr int seq_len = 256;  // 16 blocks >= 8 → split-K heuristics fire
```

Split-K goes through `crosswarp_reduce_splitk`, which carries no sink term (the sink is folded in by the split reduction kernel). `M31` mutates `crosswarp_reduce_and_write` — the **non-split** path — which no sink test reaches. These are the only two sink tests in the suite (`rg -l 'sinks' tests/`).

The file already records the same lesson one iteration earlier (`tests/test_paged_attention.cu:674`):

> The split-K branch only activates when scratch is set — none of the older tests set it, so the hd=64 split-K path was never covered.

That fix added split-K coverage and left the non-split path uncovered.

## Why the test suite missed it

Escape class: **E6** — the test exists, asserts against a real CPU reference, and its input shape routes around the mutated code.

## Severity

S1, scoped: silently wrong attention output for gpt-oss-class models (the only ones with learned sinks) whenever decode takes the non-split path — i.e. short contexts, or batch×heads large enough that the split-K heuristic declines (`compute_splitk_splits`, `attention_paged_common.cuh:260`).

## Suggested test

Add a third case to `run_gptoss_shape_splitk_case` with `seq_len` small enough that split-K does not fire (e.g. `seq_len = 48` → 3 blocks < the `num_ctx_blocks >= 4` threshold), `with_sinks=true`, against the same CPU reference. One extra `TEST(...)` line plus a parameter; the reference already handles the sink column.

Found by the test-hardening mutation baseline; see `docs/audit/MUTATION_BASELINE.md`.

**Comment by kekzl** [comment]
## Correction: the uncovered path is narrower and further away than I wrote

My original diagnosis — "the sink tests are shaped so split-K fires, so the non-split reduction is uncovered" — is only half right. I wrote a test for it and it does **not** kill the mutant, so here is what is actually true.

There are **three** independent sink implementations, not two:

| Reduction | Sink handling | Covered by |
|---|---|---|
| `paged_attention_gqa_kernel` (FP16, non-split) | inline, `attention_paged.cu:266-274` | now covered — see below |
| `paged_attention_reduce_kernel` (split-K merge) | inline, `attention_paged.cu:1087-1099` | `GQA_SplitK_HD64_Sinks` |
| `crosswarp_reduce_and_write` (shared helper) | `attention_paged_common.cuh:384-391` | **nothing** |

M31 mutates the third one. Its callers are the SM120 thread-block-**cluster** kernel (`attention_paged.cu:1342`) and every **quantised-KV** decode kernel — int4 (`:156`), int8 (`:452`), nvfp4 (`:200`), nvfp4_tc (`:605`), fp8 (`:548`).

Every test that touches those paths passes `n_sinks=0` and a null sink pointer:

```
$ rg -n 'n_sinks' tests/test_attention_paged_nvfp4_tc_residual.cu tests/test_fp8_kv_cache.cu
...:/*n_sinks=*/0,     (×4)
```

So the sink term in that helper is reached by **zero** tests, and the correct statement of the gap is: *no test performs a decode with learned sinks against a quantised KV cache or through the cluster kernel* — i.e. exactly the configuration gpt-oss runs in production when the KV cache is FP8.

## What I did add

```
PagedAttentionTest.GQA_NoSplitK_HD64_Sinks
PagedAttentionTest.GQA_NoSplitK_HD64
```

`run_gptoss_shape_splitk_case` is now parameterised by `seq_len`; at 48 (3 blocks < the `num_ctx_blocks >= 4` threshold in `compute_splitk_splits`) split-K does not fire, so the FP16 non-split kernel's own inline sink handling is exercised for the first time. Both pass on `main`, and both are new coverage — they just do not reach M31's target.

## What would actually close it

Extend `tests/test_fp8_kv_cache.cu`'s existing decode harness with a sink variant: it already builds Q/K/V, an FP8 KV cache and a CPU reference, and `paged_attention_decode_fp8` already takes `n_sinks` + the sink pointer. The reference in `run_gptoss_shape_splitk_case` (`tests/test_paged_attention.cu:700-720`) shows the expected maths — softmax over `[scores, sink]` with the sink column dropped from the numerator — and can be lifted as-is.

Severity is unchanged (S1, scoped to sink models), but the trigger is narrower than I first said: gpt-oss-class model **and** a quantised KV cache or the cluster path. `--kv-fp8` is a normal thing to run, so this is reachable in production.

**Comment by kekzl** [comment]
## Same root as this issue: `compute_splitk_splits` is only reachable from the quantised-KV launchers

Chasing a different surviving mutant (M30 — the split-K scratch-capacity guard, `attention_paged_common.cuh:268-271`) landed on exactly the seam this issue is about, so recording it here rather than opening a near-duplicate.

**The guard has no caller on the FP16 path.**

```
$ rg -n 'compute_splitk_splits' src/compute/*.cu
attention_paged_nvfp4.cu:377      attention_paged_int4.cu:535
attention_paged_nvfp4.cu:473      attention_paged_int8.cu:481
attention_paged_nvfp4_tc.cu:1080  attention_paged_fp8.cu:576
```

`paged_attention_decode` (FP16) makes its own split decision inline. So do the FP16 tests: I built a shape with a deliberately undersized scratch and a canary past the advertised size, and it cannot fail, because the code under test is never entered. Those tests were written and then **reverted** rather than committed with a premise I had already disproven.

That is the same structural fact as the sink gap in this issue: `crosswarp_reduce_and_write`, `compute_splitk_splits` and the sink term all live in `attention_paged_common.cuh`, shared by the cluster kernel and the five quantised-KV decode kernels — and **every test that touches those passes `n_sinks=0` and a default-sized scratch.** Two independent mutants now point at the same uncovered set of launchers.

### What would close both at once

`tests/test_fp8_kv_cache.cu` already builds the whole setup — Q/K/V, an FP8 KV cache, a CPU reference, and `SplitKConsistency` even allocates split-K scratch and compares against the non-split result. Two variants on that harness cover both mutants:

1. **Sinks** (this issue): pass a non-null sink pointer and `n_sinks > 0`, with the reference from `run_gptoss_shape_splitk_case` (`tests/test_paged_attention.cu:700-720`) — softmax over `[scores, sink]`, sink column dropped from the numerator.
2. **Capacity** (M30): allocate the scratch full size but advertise only one split's worth via `paged_attention_set_splitk_scratch(ptr, one_split_bytes)`, fill the remainder with a canary, and assert both that the output still matches and that the canary is intact. With the guard the launcher falls back to `num_splits=1`; without it, it writes past the advertised size.

The canary is the part that matters: `cudaMalloc` rounds up generously, so an overrun lands in slack that nothing reads and the output stays correct. Comparing outputs alone cannot see it — measured, not assumed.

**Comment by kekzl** [comment]
Closed by #1330. M31 is dead — after two earlier attempts, and both of them failed for reasons worth recording, because they are not the reason this issue originally gave.

## The recipe in this issue would not have worked either

It proposed extending `tests/test_fp8_kv_cache.cu` with a sink variant. `paged_attention_decode_fp8` has no sink parameter at all, and neither do the int4/int8/nvfp4 launchers: **six of the seven call sites of `crosswarp_reduce_and_write` leave `attn_sinks` at its `nullptr` default.** Exactly one passes a sink pointer — the SM120 thread-block-cluster kernel (`attention_paged.cu:1342`) — and it runs only when

- there is no split-K scratch, so `num_splits` stays 1, **and**
- `n_q_per_kv` ∈ {2,4,8}, `head_dim` ∈ {64,96,128,256}, **and**
- there are ≥ 8 context blocks.

`GQA_NoSplitK_HD64_Sinks` (added for this issue) misses it because 48 tokens is 3 blocks — under the 8-block floor, so the plain GQA kernel runs, which handles sinks inline. `GQA_SplitK_HD64_Sinks` misses it because 256 tokens *with* scratch goes split-K, whose reduction also handles sinks inline. Three sink implementations, and the tests covered the other two.

## A second defect had to be fixed before any test could fail

The sink logit competes against the *sum* of `seq_len` score exponentials, so a value inside the score range contributes almost nothing:

| sinks | share of softmax mass | max ∣Δ output∣ when the term is dropped |
|---|---|---|
| `-1.0 + 0.05h` (as written) | 0 – 12.6 % | **0.00167** |
| `+4.0 + 0.05h` (now) | 0.5 – 95.6 % | 0.0566 |

against `EXPECT_NEAR(..., 0.05f)`. **The signal was 30x smaller than the tolerance.** The tolerance was itself unmoored: measured max error against the fp32 host reference is 1.7e-4 (seq_len 48) and 7.1e-5 (256), so `0.05` was ~300x looser than the kernel needs. Now `2e-3` — 12x over the kernel error, 4.7x-28x under the sink signal.

## Evidence

```
$ ./build-dev/test-attention --gtest_filter='PagedAttentionTest.*'    # M31 injected
[  FAILED  ] PagedAttentionTest.GQA_Cluster_HD64_Sinks
[  PASSED  ] 15 tests.
```

One failure, and it is the new case — which doubles as the proof of the reachability analysis, since both older sink tests stay green under the same mutant. Clean tree: `test-attention` 203/203.

Deliberately **not** the mutation harness's own verdict: its re-run reported `KILLED` via `test-e2e`, listing `TokenizeRoundtrip` among the detecting tests. A tokenizer round-trip cannot observe a softmax denominator. That lane's baseline is red for environmental reasons (the documented WSL2/WDDM peak-VRAM behaviour), and it had already produced two false *survivor* verdicts in iteration 1; this time it produced a false kill.

Write-up: `docs/audit/TEST_HARDENING_LOG.md`, iteration 13.
